presto-raptor/src/main/java/com/facebook/presto/raptor/storage/BucketBalancer.java

```diff
@@ -0,0 +1,369 @@
+                        break;
+                    }
+
+                    // identify nodes with bucket counts lower than the computed target, and greedily select from this set based on projected disk utilization.
+                    // greediness means that this may produce decidedly non-optimal results if one looks at the global distribution of buckets->nodes.
+                    // also, this assumes that nodes in a cluster have identical storage capacity
+                    String target = activeNodes.stream()
+                            .filter(candidate -> !candidate.equals(source) && allocationCounts.count(candidate) < targetMax)
+                            .sorted((a, b) -> Integer.compare(allocationCounts.count(a), allocationCounts.count(b)))
+                            .min((a, b) -> Double.compare(allocationBytes.get(a), allocationBytes.get(b)))
+                            .orElseThrow(() -> new RuntimeException("unable to find target for rebalancing"));
+
+                    long bucketSize = clusterState.getDistributionBucketSize().get(distribution);
+
+                    // don't move bucket if it doesn't reduce imbalance
+                    if (activeNodes.contains(source) && (allocationCounts.count(source) == targetMax && allocationCounts.count(target) == targetMin)) {
+                        break;
+                    }
+
+                    allocationCounts.remove(source);
+                    allocationCounts.add(target);
+                    allocationBytes.compute(source, (k, v) -> v - bucketSize);
+                    allocationBytes.compute(target, (k, v) -> v + bucketSize);
+
+                    sourceToAllocationChanges.put(
+                            existingAssignment.getNodeIdentifier(),
+                            new BucketAssignment(existingAssignment.getDistributionId(), existingAssignment.getBucketNumber(), target));
+                }
+            }
+        }
+
+        return sourceToAllocationChanges;
+    }
+
+    private int updateAssignments(Multimap<String, BucketAssignment> sourceToAllocationChanges)
+    {
+        // perform distributionAssignments in decreasing order of source node load
+        List<String> sourceNodes = sourceToAllocationChanges.asMap().entrySet().stream()
+                .sorted((a, b) -> Integer.compare(b.getValue().size(), a.getValue().size()))
+                .map(Map.Entry::getKey)
+                .collect(toList());
+
+        int moves = 0;
+        for (String source : sourceNodes) {
+            for (BucketAssignment reassignment : sourceToAllocationChanges.get(source)) {
+                // todo: rate-limit new distributionAssignments
+                shardManager.updateBucketAssignment(reassignment.getDistributionId(), reassignment.getBucketNumber(), reassignment.getNodeIdentifier());
+                bucketsBalanced.update(1);
+                moves++;
+                log.info("Distribution %s: Moved bucket %d from %s to %s",
```

Comment:
Use `%s` instead of `%d` since it's safer (works with any type). Only use `%d` when you need specific formatting like `%02d`